### ISSUE TITLE
Receive: poor memory efficiency when (un)marshaling data

### ISSUE THREAD

**Issue by fpetkovski** [thanos-io/thanos#5751]
### Is your proposal related to a problem?

I took a profile of a bloated Thanos Receiver, and noticed that a huge amount of memory is being allocated for just marshaling and unmarshaling data. This is a big surprise as I would expect the majority of the memory usage would come from TSDBs.

### Describe the solution you'd like

We should investigate why marshaling data has such a significant impact on memory. 

### Describe alternatives you've considered

N/A

### Additional context

Here is a link to the profile: https://pprof.me/ccc9a28/

<img width="1415" alt="image" src="https://user-images.githubusercontent.com/1286231/193525950-34ad4dcc-d5ed-4a54-9987-9d8fb572550e.png">

**Comment by bwplotka** [comment]
Agree, thanks! 💪🏽 We see similar on our side I think @matej-g @PhilipGough

**Comment by douglascamata** [comment]
@fpetkovski: Why would TSDBs allocate more memory than marshalling? Aren't TSDBs just pointing to memory previously allocated by the marshalling? 🤔

**Comment by fpetkovski** [comment]
TSDBs hold memory for a longer time period, and I assume we don't marshal everything from them during series requests.

**Comment by matej-g** [comment]
I think we've seen similar numbers in our profiles, for us the biggest issue was switch to snappy compression on `0.28.0`, as I was mentioning in https://github.com/thanos-io/thanos/pull/5575#issuecomment-1268095662, which after we switched back to no compression we got back to numbers similar to `0.27.0`.  @PhilipGough has now run a couple more tests comparing with / without compression and the issue seems to be especially pronounced on instances where we ingest big requests. @fpetkovski @PhilipGough feel free to add more from your observations.

But speaking solely on (un)marshalling, yes, that could be the next thing to look at. Although I wonder where / what to start with :thinking:.

**Comment by douglascamata** [comment]
One of the question I have in mind at the moment is: is marshalling's high in-use memory numbers due to few huge gRPC messages? Does it get better in the scenario with more numerous but smaller messages?

**Comment by matej-g** [comment]
Another related issue, I'm wondering what kind of improvement we'd see if / when we move to the Vitess framework https://github.com/thanos-io/thanos/issues/4557

**Comment by fpetkovski** [comment]
I was going to suggest the same, we might be able to start pooling requests once that is merged.

**Comment by fpetkovski** [comment]
Could be related to https://github.com/klauspost/compress/issues/442

**Comment by matej-g** [comment]
@juanrh definitely! Could you open (even if a draft) PR with the proposed changes? We could discuss it in full there and perhaps look at some benchmarks.

**Comment by juanrh** [comment]
@matej-g I just sent https://github.com/thanos-io/thanos/pull/5795 for that
